src/Spotify/SpotifyBackend.cpp

```diff
@@ -10,39 +10,247 @@
+  if (returnValue.has_value()) {                                             \
+    auto ret = errorHandler(returnValue.value());                            \
+    if (ret.has_value()) {                                                   \
+      LOG(ERROR) << ret.value().getErrorMessage() << std::endl;              \
+      return ret.value();                                                    \
+    } else {                                                                 \
+      tokenString = mSpotifyAuth.getAccessToken();                           \
+      returnValue = functionCall;                                            \
+                                                                             \
+      if (returnValue.has_value()) {                                         \
+        LOG(ERROR) << returnValue.value().getErrorMessage() << std::endl;    \
+                                                                             \
+        return returnValue.value();                                          \
+      }                                                                      \
+    }                                                                        \
+  }
+
+SpotifyBackend::~SpotifyBackend() {
+  mSpotifyAuth.stopServer();
+}
 
 TResultOpt SpotifyBackend::initBackend() {
-  return nullopt;
+  // start server for authentication
+  auto startServerRet = mSpotifyAuth.startServer();
+
+  return startServerRet;
 }
 
-TResult<vector<BaseTrack>> SpotifyBackend::queryTracks(string const &pattern,
-                                                       size_t const num) {
-  auto tmp = vector<BaseTrack>();
-  tmp.emplace_back();
-  return tmp;
+TResult<std::vector<BaseTrack>> SpotifyBackend::queryTracks(
+    std::string const &pattern, size_t const num) {
+  std::string token = mSpotifyAuth.getAccessToken();
+
+  TResult<SpotifyPaging> retVal;
+  SPOTIFYCALL_WITH_REFRESH(
+      retVal, mSpotifyAPI.search(token, pattern, QueryType::track, num), token);
+
+  auto page = std::get<SpotifyPaging>(retVal);
+  std::vector<BaseTrack> tracks;
+
+  for (auto const &elem : page.getTracks()) {
+    BaseTrack track;
+    track.artist = "";
+    track.iconUri = "";
+
+    track.title = elem.getName();
+    track.album = elem.getAlbum().getName();
+    track.duration = elem.getDuration();
+    track.trackId = elem.getUri();
+
+    if (!elem.getAlbum().getImages().empty()) {
+      track.iconUri = elem.getAlbum()
+                          .getImages()[0]
+                          .getUrl();  // on first place is the biggest one
+    }
+
+    for (auto &artist : elem.getArtists()) {
+      track.artist += artist.getName() + " & ";
+    }
+    std::string test;
+    track.artist.erase(track.artist.find_last_of(" & "), 3);
+
+    tracks.emplace_back(track);
+  }
+
+  return std::move(tracks);
 }
 
 TResultOpt SpotifyBackend::setPlayback(BaseTrack const &track) {
```

Comment:
at this point we probably do not have a full `BaseTrack` object. could you change the parameter type to `TTrackID`? it seems like the method doesnt need more information anyway

Reply:
sure, can do it, but im wondering why we dont have a full base track? thought in the queue is a QueueTrack, wich is a BaseTrack

Reply:
hmm thats actually true 🤔 
so i guess it doesnt matter 🤷‍♂